get-oui.1
.\" Copyright (C) Roy Hills
.\"
.\" Copying and distribution of this file, with or without modification,
.\" are permitted in any medium without royalty provided the copyright
.\" notice and this notice are preserved.
.\"
.TH GET-OUI 1 "October 28, 2022"
.\" Please adjust this date whenever revising the man page.
.SH NAME
get-oui \- Fetch the arp-scan OUI file from the IEEE website
.SH SYNOPSIS
.B get-oui
.RI [ options ]
.SH DESCRIPTION
.B get-oui
fetches the Ethernet OUI file from the IEEE website, and saves it in the
format used by arp-scan.
.PP
The OUI file contains all of the OUIs (Organizationally Unique Identifiers)
that have been registered with IEEE.  Each OUI entry in the file specifies
the first 24-bits of the 48-bit Ethernet hardware address, leaving the
remaining 24-bits for use by the registering organisation.  For example
the OUI entry "080020", registered to Sun Microsystems, applies to any
Ethernet hardware address from
.I 08:00:20:00:00:00
to
.I 08:00:20:ff:ff:ff
inclusive.  Each OUI assignment represents a total of 2^24 (16,777,216)
Ethernet addresses.
.PP
Every major Ethernet hardware vendor registers an OUI for their equipment,
and larger vendors will need to register more than one.  For example, 3Com
have a total of 37 OUI entries.  Organisations that only produce a small
number of Ethernet devices will often obtain an IAB registration instead.
See
.BR get-iab (1)
for details.
.PP
This script can be used to update the
.B arp-scan
OUI file from the latest data on the IEEE website.  Most of the Ethernet
addresses in use belong to an OUI registration, so this is the most
important of the files that
.B arp-scan
uses to decode Ethernet hardware addresses.  You should therefore run
.B get-oui
occasionally to keep the
.B arp-scan
OUI file up to date.
.PP
The OUI data is fetched from the URL
.I http://standards-oui.ieee.org/oui/oui.txt
and the output file is saved to the file
.I ieee-oui.txt
in the current directory. The URL to fetch the data from can be changed with
the
.B -u
option, and the output file name can be changed with the
.B -f
option.
.PP
The
.I ieee-oui.txt
file that is produced by this script is used by
.B arp-scan
to determine the Ethernet card vendor from its hardware address.
.PP
The directory that
.B arp-scan
will look for the
.I ieee-oui.txt
file depends on the options used when it was built.  If it was
built using the default options, then it will look in
.IR /usr/local/share/arp-scan .
.SH OPTIONS
.TP
.B -h
Display a brief usage message and exit.
.TP
\fB-f \fI<fn>\fR
Write the output to the specified file instead of the default
.I ieee-oui.txt.
.TP
\fB-u \fI<URL>\fR
Use the specified URL to fetch the raw OUI data from instead of the default
.I http://standards-oui.ieee.org/oui/oui.txt.
.TP
.B -v
Display verbose progress messages.
.SH FILES
.TP
.I ieee-oui.txt
The default output file.
.SH EXAMPLES
.nf
$ ./get-oui -v
Renaming ieee-oui.txt to ieee-oui.txt.bak
Fetching OUI data from http://standards-oui.ieee.org/oui/oui.txt
Fetched 5293585 bytes
Opening output file ieee-oui.txt
32812 OUI entries written to file ieee-oui.txt
.fi
.SH NOTES
.B get-oui
is implemented in Perl, so you need to have the Perl interpreter installed on
your system to use it.
.PP
.B get-oui
uses the
.I LWP::UserAgent
Perl module to fetch the data from the IEEE website. You must have this module
installed on your system for
it to work. This module is available on most distributions, often called
.IR libwww-perl .
It is also available in source form from CPAN.
.PP
You can use a proxy server by defining the
.I http_proxy
environment variable.
.SH "SEE ALSO"
.TP
.BR arp-scan (1)
.TP
.BR get-iab (1)
.TP
.BR arp-fingerprint (1)


NEWS
This file gives a brief overview of the major changes between each arp-scan
release.  For more details please read the ChangeLog file.

YYYY-MM-DD arp-scan 1.9.9 (in progress)

* New Features:

  - Added POSIX.1e capabilities support for systems with libcap (Linux only)
  - New --format option provides flexible output format (see manpage or help)
  - get-oui script now stores IEEE registries MA-L (OUI), MA-M, MA-S (OUI36)
    and IAB in the single file ieee-oui.txt. The ieee-iab.txt file and the
    --iabfiles option to arp-scan have been removed as they are no longer
    required.

2021-10-08 arp-scan 1.9.8 (git tag 1.9.8)

* New Features:

  - Allow the use of Linux IP aliases such as eth0:0 for the interface name.
  - Permit regular MAC addresses e.g. 00:0c:29:b9:43:1b in mac-vendor.txt.
  - --limit=n option exits after n of hosts have responded, exit 1 for <n
  - --resolve option to resolve responding IP addresses to hostnames

* Fixed bugs:

  - Potential buffer overrun in unmarshal_arp_pkt().
  - arp-scan aborts with EAGAIN on busy network or using high bandwidth
  - late ARP responses could sometimes be incorrectly flagged as duplicate

* General improvements:

  - Updated IEEE URLs in download perl scripts.
  - Updated source for Mersenne RNG and replacement strlcat/strlcpy & getopt.
  - Updated for compatability with autoconf 2.71
  - "make distcheck" works now
  - Number of responding hosts reported no longer counts duplicate packets.
  - Many typos corrected and edge cases fixed.

* Misc Changes:

  - CI framework migrated from travis-ci to github actions.
  - Several new tests for "make check"

2019-11-10 arp-scan 1.9.7

* Improved error messages from libpcap functions.

* Remove obsolescent and unused autoconf macros.  arp-scan 1.9.7 assumes that
  the C compiler is ANSI C (C89) compliant.

2019-10-13 arp-scan 1.9.6

* Use libpcap function pcap_set_immediate_mode() instead of ioctl calls to
  ensure packets are delivered immediately. This fixes the bug which caused
  arp-scan on linux to not report any hosts with libpcap 1.9.1. This change
  means arp-scan now requires libpcap 1.5.0 or later and will not work with
  earlier versions.

* Fix compiler warnings caused by the depreciated function pcap_lookupdev()
  in libpcap 1.9.0 and later.

2016-09-03 arp-scan 1.9.5

* Use posix hash table functions hcreate(), hsearch() and hdestroy() instead
  of the gas hash table code. Thanks to nihilus for the suggestion.

* Remove function replacement for inet_aton, as this was only required for
  Solaris 8, which is now considered obsolete.

* Added arp-fingerprint patterns for FreeBSD 10.3, DragonflyBSD 4.6, Windows10,
  Linux 4.0, Linux 4.6, OpenBSD 5.9, NetBSD 7.0.

* Added "-l" option to arp-fingerprint to support fingerprinting all hosts on
  the local network. Thanks to Rhig for the pull request.

* Use the source_mac rather than interface_mac in the pcap filter, to permit
  reception of packets with spoofed MAC source address. Thanks to tissieres
  for the pull request.

* Use the libpcap 1.0 API functions pcap_create() instead of pcap_open_live().
  This means that arp-scan now requires libpcap 1.0 or later and will not work
  with earlier libpcap versions.

* Updated IEEE OUI and IAB download locations to reflect IEEE website changes.

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 22487 OUI entries
  and 4575 IAB entries.

2013-11-24 arp-scan 1.9.2

* Added new --plain (-x) option to suppress printing of header and footer text,
  and only display one output line for each responding host. Idea from Stefan
  Tomanek's arp-scan fork on github at https://github.com/wertarbyte/arp-scan.

* Use LWP::UserAgent instead of LWP::Simple in get-oui and get-iab to enable
  the raw content to be obtained, which avoids Unicode/UTF-8 issues.

* Added arp-fingerprint patterns for WIZnet W5100 and Cisco IOS 15.0.

* Moved arp-scan development from internal SVN repository to github at
  https://github.com/royhills/arp-scan. The move to git means that the commit
  object names are now SHA1 hashes instead of increasing integer values. So
  they are longer suitable for internal file versions with rcsid variables.
  Accordingly the rcsid variables have been removed.

2013-07-24 arp-scan 1.9:

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 18157 OUI entries
  and 4414 IAB entries.

* Use autoconf 2.69 and automake 1.11 to add support for ARM 64-bit CPUs.

* Use libpcap functions to obtain the interface IP address and send the ARP
  packet, instead of using our own link-layer specific functions. The only
  link-layer specific function that we still need is get_hardware_address()
  to obtain the interface MAC address. This means we now require libpcap 0.9.3
  or later.

* Added support for Dragonfly BSD.

* The -u option to get-iab and get-oui scripts now works.

* get-oui and get-iab scripts now get the OUI and IAB files from the new
  locations on the IEEE website, and allow whitespace at the beginning of
  the line.

* If the MAC/Vendor file locations are not explicitly specified, look for them
  in the current directory and then in their default location.

* Raised default timeout from 100ms to 500ms.

* Added new --rtt (-D) option to display the packet round-trip time.

* Include <net/bpf.h> header file early in link-bpf.c to avoid BPF symbol
  problems on some BSD based operating systems.

* Added arp-fingerprint patterns for GNU/Hurd, Amazon Kindle (Linux 2.6),
  BeOS, Windows 8, Recent Linux, FreeBSD, NetBSD and OpenBSD versions, and
  RiscOS.

* Added data file "pkt-custom-request-vlan-llc.dat" to the tarball to allow
  the ARP request packet generation self test to complete successfully.

* Various minor bug fixes and improvements.

2011-03-01 arp-scan 1.8:

* Updated IEEE OUI and IAB MAC/Vendor files.  There are now 14707 OUI entries
  and 3542 IAB entries.

* Added support for trailer ARP replies, which were used in early versions
  of BSD Unix on VAX.


get-oui.pl
#!/usr/bin/perl
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# $Id$
#
# get-oui.pl -- Fetch the OUI file from the IEEE website
#
# Author: Roy Hills
# Date: 16 March 2006
#
# This script downloads the Ethernet OUI file from the IEEE website, and
# converts it to the format needed by arp-scan.
#
use warnings;
use strict;
use Getopt::Std;
use LWP::Simple;
#
my $default_url = 'http://standards.ieee.org/regauth/oui/oui.txt';
my $default_filename='oui.txt';
#
my $usage =
qq/Usage: get-oui.pl [options]
Fetch the Ethernet OUI file from the IEEE website, and save it in the format
used by arp-scan.

'options' is one or more of:
        -h Display this usage message.
        -f FILE Specify the output OUI file. Default=$default_filename
        -u URL Specify the URL to fetch the OUI data from.
           Default=$default_url
        -v Give verbose progress messages.
/;
my %opts;
my $verbose;
my $filename;
my $url;
my $lineno = 0;
#
# Process options
#
die "$usage\n" unless getopts('hf:v',\%opts);
if ($opts{h}) {
   print "$usage\n";
   exit(0);
}
if (defined $opts{f}) {
   $filename=$opts{f};
} else {
   $filename=$default_filename;
}
if (defined $opts{u}) {
   $url=$opts{u};
} else {
   $url=$default_url;
}
$verbose=$opts{v} ? 1 : 0;
#
# If the output filename already exists, rename it to filename.bak before
# we create the new output file.
#
if (-f $filename) {
   print "Renaming $filename to $filename.bak\n" if $verbose;
   rename $filename, "$filename.bak" || die "Could not rename $filename to $filename.bak\n";
}
#
# Fetch the content from the URL
#
print "Fetching OUI data from $url\n" if $verbose;
my $content = get $url;
die "Could not get OUI data from $url\n" unless defined $content;
my $content_length = length($content);
die "Zero-sized response from from $url\n" unless ($content_length > 0);
print "Fetched $content_length bytes\n" if $verbose;
#
# Open the output file for writing.
#
print "Opening output file $filename\n" if $verbose;
open OUTPUT, ">$filename" || die "Could not open $filename for writing";
#
# Write the header comments to the output file.
#
my ($sec,$min,$hour,$mday,$mon,$year,undef,undef,undef) = localtime();
$year += 1900;
$mon++;
my $date_string = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year, $mon, $mday,
                          $hour, $min, $sec);
my $header_comments =
qq/# oui.txt -- Ethernet vendor OUI file for arp-scan
#
# This file contains the Ethernet vendor OUIs for arp-scan.  These are used
# to determine the vendor for a give Ethernet interface given the MAC address.
#
# Each line of this file contains an OUI-vendor mapping in the form:
#
# <OUI><TAB><Vendor>
#
# Where <OUI> is the first three bytes of the MAC address in hex, and <Vendor>
# is the name of the vendor.
#
# Blank lines and lines beginning with "#" are ignored.
#
# This file was automatically generated by get-oui.pl at $date_string
# using data from $url
#
/;
print OUTPUT $header_comments;
#
# Parse the content received from the URL, and write the OUI entries to the
# output file.  Match lines that look like this:
# 00-00-00   (hex)                XEROX CORPORATION
# and write them to the output file looking like this:
# 000000	XEROX CORPORATION
#
while ($content =~ m/^(\w+)-(\w+)-(\w+)\s+\(hex\)\s+(.*)$/gm) {
   print OUTPUT "$1$2$3\t$4\n";
   $lineno++;
}
#
# All done.  Close the output file and print OUI entry count
#
close OUTPUT || die "Error closing output file\n";
print "$lineno OUI entries written to file $filename\n" if $verbose;
